drivers/sdhc/sdhci_common.c

```diff
@@ -428,6 +428,72 @@ static int sdhci_transfer_data(struct sdhci_common *sdhci_ctx, struct sdhc_data
 	return sdhci_poll_data_complete(sdhci_ctx, timeout / 100);
 }
 
+static void sdhc_adma_desc_fill(struct sdhc_adma_desc *desc, uintptr_t dma_addr, uint16_t len,
+				bool end)
+{
+	uint16_t cmd;
+
+	cmd = SDHC_ADMA_DESC_VALID | SDHC_ADMA_DESC_ACT_TRAN;
+	if (end) {
+		cmd |= SDHC_ADMA_DESC_END;
+	}
+
+	desc->cmd = cmd;
+	desc->len = len;
+	desc->addr = dma_addr;
+	LOG_DBG("sdhc:dma_desc: %016llx", *((uint64_t *)desc));
+}
+
+static void sdhc_prepare_adma_table(struct sdhci_common *sdhci_ctx, struct sdhc_data *data,
+				    uintptr_t dma_addr)
+{
+	uint32_t trans_bytes = data->block_size * data->blocks;
+	uint32_t desc_count = DIV_ROUND_UP(trans_bytes, SDHC_ADMA_DESC_MAX_LEN);
+	struct sdhc_adma_desc *desc = sdhci_ctx->adma_descs;
+
+	if (desc_count >= sdhci_ctx->adma_descs_num) {
+		LOG_ERR("sdhc:adma: xfer desc number (%u) overflow, max:%u ", desc_count,
+			sdhci_ctx->adma_descs_num);
+	}
+
+	while (--desc_count) {
+		sdhc_adma_desc_fill(desc, dma_addr, SDHC_ADMA_DESC_MAX_LEN, false);
+		dma_addr += SDHC_ADMA_DESC_MAX_LEN;
+		trans_bytes -= SDHC_ADMA_DESC_MAX_LEN;
+		desc++;
+	}
+
+	sdhc_adma_desc_fill(desc, dma_addr, trans_bytes, true);
+
+#if !defined(CONFIG_NOCACHE_MEMORY)
+	sys_cache_data_flush_range(sdhci_ctx->adma_descs,
+				   (sdhci_ctx->adma_descs_num * sizeof(struct sdhc_adma_desc)));
+#endif
+
+	sys_write32((uintptr_t)sdhci_ctx->adma_descs, sdhci_ctx->reg_base + SDHCI_ADMA_ADDRESS);
```

Comment:
Shouldn't you use the PHYS address here for DMA descriptors?
Of course, if you have direct mapping of virtual to physical addresses, you shouldn't, but is that the case?